component/prometheus/fanout.go

```diff
@@ -0,0 +1,131 @@
+package prometheus
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/prometheus/prometheus/model/exemplar"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/model/metadata"
+
+	"github.com/prometheus/prometheus/storage"
+)
+
+var _ storage.Appendable = (*Fanout)(nil)
+
+type intercept func(ref storage.SeriesRef, l labels.Labels, t int64, v float64) (storage.SeriesRef, labels.Labels, int64, float64, error)
+
+// Fanout supports the default Flow style of appendables since it can go to multiple outputs. It also allows the intercepting of appends.
+type Fanout struct {
+	mut sync.RWMutex
+	// intercept allows one to intercept the series before it fans out to make any changes. If labels.Labels returns nil the series is not propagated.
+	// Intercept shouuld be thread safe and can be called across appenders.
+	intercept func(ref storage.SeriesRef, l labels.Labels, t int64, v float64) (storage.SeriesRef, labels.Labels, int64, float64, error)
+
+	// children is where to fan out.
+	children []storage.Appendable
+
+	// ComponentID is what component this belongs to.
+	componentID string
+}
+
+// NewFanout creates a fanout appendable.
+func NewFanout(inter intercept, children []storage.Appendable, componentID string) *Fanout {
+	return &Fanout{
+		intercept:   inter,
+		children:    children,
+		componentID: componentID,
+	}
+}
+
+// UpdateChildren allows changing of the children of the fanout.
+func (f *Fanout) UpdateChildren(children []storage.Appendable) {
+	f.mut.Lock()
+	defer f.mut.Unlock()
+	f.children = children
+}
+
+// Appender satisfies the Appendable interface.
+func (f *Fanout) Appender(ctx context.Context) storage.Appender {
+	f.mut.RLock()
+	defer f.mut.RUnlock()
+
+	app := &appender{
+		children:    make([]storage.Appender, len(f.children)),
+		intercept:   f.intercept,
+		componentID: f.componentID,
+	}
+	for i, x := range f.children {
+		if x == nil {
+			continue
+		}
+		app.children[i] = x.Appender(ctx)
+	}
```

Comment:
What about setting the capacity of `children` instead of the length? That way you can remove all the nil checks in the appender itself and guarantee that each child is non-nil.

Reply:
Removed the nil checks from the appender and simplified a bit of the logic. Nils are allowed in the appendable but not the appender.